### ISSUE TITLE
Large Discrepancy in Results When Running KdV Equation on Different Hardware (T4 vs. TPU v2-8) in Colab

### ISSUE THREAD

**Issue by chang-change** [PredictiveIntelligenceLab/jaxpi#22]
Hello,

I’m experiencing a significant difference in the results when running the Korteweg-de Vries (KdV) equation on Google Colab using different hardware setups (T4 GPU versus TPU v2-8) and on workstation with 3090 GPU. The same code produces considerably different outputs depending on which hardware is used.

I would like to understand why this discrepancy occurs and if there are any recommendations for achieving consistent results across different hardware configurations.

Any insights or suggestions would be greatly appreciated!

Thank you.
![image](https://github.com/user-attachments/assets/9792c11f-37e4-4405-8889-b6755e71e746)

**Comment by sifanexisted** [comment]
Thanks for bringing this to our attention. We’re encountering the same issue across different hardware setups, CUDA, cuDNN, and JAX versions. It seems to be related to the default floating-point precision for matrix operations. Please try the following command to see if it helps (maybe lower precision gives better accuracy):

`jax.config.update("jax_default_matmul_precision", "highest")`

**Comment by chang-change** [comment]
Thank you for your response. Where should I place this code, and in which file？

**Comment by sifanexisted** [comment]
Sorry for the late reply. You may want to put this line of code in the beginning of main.py

**Comment by chang-change** [comment]
Thank you! I will try.
